### PR TITLE
Added AODNet Image Dehazing Example

### PR DESCRIPTION
**Pull request by soumik12345**
Image Dehazing is an important problem in Computer Vision. It can not only be used to enhance photographs clicked in hazy conditions, but also can be used to augment other computer vision systems such as Object Detection and Semantic Segmentation as a pre-processing step. In this example, we implement [AODNet](https://arxiv.org/abs/1707.06543v1), a simple and lightweight Deep CNN model that directly generates a clean image from a hazy image. The end-to-end design of AODNet makes it easy to embed AOD-Net into other deep models such as Faster-RCNN and YOLO as a pre-processing step, making it easier for such a model to detect and recognize objects from a clean image.

### CODE REVIEW

**Comment by soumik12345** [comment]
@fchollet I would continue to refactor the code for Image Dehazing as per your suggestions in [https://github.com/keras-team/keras-io/pull/440](https://github.com/keras-team/keras-io/pull/440) since I accidentally deleted the fork of this [https://github.com/keras-team/keras-io/pull/439](https://github.com/keras-team/keras-io/pull/439).

File: examples/vision/aodnet.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+"""shell
+!nvidia-smi -L
+"""
+
+import os
+import gdown
+import numpy as np
+from PIL import Image
+from glob import glob
+from glob import glob
```
Comment (fchollet) [inline examples/vision/aodnet.py:30]: Please apply fixes to all comments I made on the previous PR?

**Comment by soumik12345** [comment]
- [x] I kept the `!nvidia-smi -L`, I think it would be better for reproducibility if we know which GPU the demo was trained on.
- [x] Removed redundant import line.
- [x] Refactored markdown links on a single line.
- [x] Replaced `os.system` with shell commands.
- [ ] Replace `get_image_file_list` using `image_dataset_from_directory`.
- [ ] Replace augmentations with preprocessing layer
- [x] Replaced the dataloader class with a stateless function.
- [x] Removed specifications of default parameters in the model.
- [x] Replaced `tf.initializers` with corresponding `tf.keras.initializers`.
- [x] Replaced `tf.math.multiply(k, inputs)` with `k * inputs`.
- [x] Removed unnecessary use of `OneDeviceStrategy`.

File: examples/vision/aodnet.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+Date created: 2021/04/10
+Last modified: 2021/04/10
+Description: Train a lightweight Deep CNN for image denhazing.
+"""
+"""
+# Introduction
+
+Image Dehazing is an important problem is Computer Vision. It can not only be used to
+enhance photographs clicked in hazy conditions, but also can be used to augment other
+computer vision systems such as Object Detection and Seantic Segmentation as a
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:13]: Suggestion:

- Use sentence case (insteda, we could add some kind of _emphasis_ on non-commonly capitalized words)
- Fix repetition ("can be used to" x 2)

```suggestion
_Image dehazing_ is an important problem is computer vision that can be used not 
only to enhance photographs clicked in hazy conditions, but also can be used to
augment other computer vision systems, such as _object detection_ and _seantic
segmentation as a
```

File: examples/vision/aodnet.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+pre-processing step. In this example, we implement
+[AODNet](https://arxiv.org/abs/1707.06543v1), a simple and light-weight Deep CNN model
+that directly generates a clean image from a hazy image. The end-to-end design of AODNet
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:16]: Suggestion:
- Clarify the abbreviation ("All-in-One Network (AOD-Net)")
- Use "AOD-Net" with a dash (as per https://arxiv.org/abs/1707.06543v1)
- Use a dash to provide a defition
- "Can generate X directly from Y" could be better
- "...deep CNN" - non capitalized
- "the AOD-Net" with the definite article ("the ... network")

```suggestion
the [All-in-One Network (AODNet)](https://arxiv.org/abs/1707.06543v1) — a simple and
lightweight deep convolutional neural network model that can generate a clean
image directly from a hazy image. The end-to-end design of the AOD-Net
```

File: examples/vision/aodnet.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+makes it easy to embed AOD-Net into other deep models such as Faster-RCNN and YOLO as a
+pre-processing step, making it easier for such model to detect and recognize objects from
+a clean image.
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:19]: Suggestion:

- Provide links to papers (YOLO, Faster-RCNN)
- Fix repetition ("make it easy" x 2) with "enable"

```suggestion
enables embedding the network into other neural network models for, such as
[Faster-RCNN](https://arxiv.org/abs/1506.01497v3) and
[YOLO](https://pjreddie.com/media/files/papers/yolo.pdf), as a pre-processing
step, making it easier for such model to detect and recognize objects from
clean images.
```

File: examples/vision/aodnet.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
+"""
+Title: An All-in-One Network for Dehazing and Beyond
+Author: [Soumik Rakshit](https://github.com/soumik12345)
+Date created: 2021/04/10
+Last modified: 2021/04/10
+Description: Train a lightweight Deep CNN for image denhazing.
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:6]: Suggestion:

- Use sentence case, refactor

```suggestion
Description: Dehaze images with a lightweight deep convolutional neural network.
```

("AOD-Net" is defined later in the intro)

File: examples/vision/aodnet.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+import os
+import gdown
+import numpy as np
+from PIL import Image
+from glob import glob
+import tensorflow as tf
+import matplotlib.pyplot as plt
+
+"""
+# Downloading Dataset
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:35]: Suggestion:

```suggestion
# Load the NYU2 Depth Database
```

(similar to https://keras.io/examples/vision/image_classification_from_scratch/)

File: examples/vision/aodnet.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+
+The authors of the AODNet paper have created synthesized hazy images using the
+groundtruth images with depth meta-data from the indoor 
+[NYU2 Depth Database](https://cs.nyu.edu/~silberman/datasets/nyu_depth_v2.html). We will
+download this dataset from Google Drive using `gdown`.
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:40]: Suggestion:

```suggestion
The experiment in the AOD-Net paper used synthesized hazy images from the
ground-truth images with depth meta-data from the indoor 
[NYU2 Depth Database](https://cs.nyu.edu/~silberman/datasets/nyu_depth_v2.html).
We will download this dataset from Google Drive using `gdown`.
```

File: examples/vision/aodnet.py [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+    quiet=False,
+)
+
+"""shell
+!unzip -qq Dehaze-NYU.zip
+!rm ./Dehaze-NYU.zip
+"""
+
+"""
+## Data Loader
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:55]: Suggestion:

```suggestion
## Define the data transformation and augmentation functions
```

File: examples/vision/aodnet.py [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+    val_split=0.1,
+)
+print(train_dataset)
+print(val_dataset)
+
+"""
+## Building AODNet Model
+
+Now we will build AODNet a subclass of  `tf.keras.Model`. The following diagrams taken
+from the paper summarizes the architecture of the AODNet model.
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:189]: Suggestion:

- Nit (grammar, etc)

```suggestion
## Define the AOD-Net model

Now we will define the AOD-Net model as a subclass of  `tf.keras.Model`. The
following diagrams from the original paper summarize its architecture:
```

File: examples/vision/aodnet.py [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+
+### The AODNet Model
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:191]:
```suggestion
### The AOD-Net model
```

File: examples/vision/aodnet.py [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+![](https://i.imgur.com/nmnF0cY.png)
+
+### K-estimation Model
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:194]:
```suggestion
### The K-estimation model
```

File: examples/vision/aodnet.py [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+
+"""
+Since we are training the model using a single GPU, i.e, Nvidia Tesla P100, we would use
+`tf.distribute.OneDeviceStrategy`, using this strategy will place any variables created
+in its scope on the specified device. We would define and compile our model in the scope
+of our distribution strategy. We would be using Adam as out optimizer, Mean Squared Error
+as the loss function and PSNR (Peak Signal Noise Ratio) as our metric.
+
+**Note:** For training AODNet in a multi-GPU environment,
+`tf.distribute.MirroredStrategy` can be used as a distribution stratgey.
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:266]: Suggestion:

- Add emphasis to "TensorFlow distribution strategy" (with this guide x 2: https://www.tensorflow.org/guide/distributed_training)
- Use the definite article "the"
- Fix a typo ("strategy")

```suggestion
In this example, we are training the model using a single GPU (an NVIDIA Tesla
P100) and the`tf.distribute.OneDeviceStrategy`. This
[TensorFlow distribution
strategy](https://www.tensorflow.org/tutorials/distribute/keras)
will place any variables created in its scope on the specified device. We will
define and compile our model in the scope of our distribution strategy, using
Adam as the optimizer, the Mean Squared Error
as the loss function, and the PSNR (Peak Signal Noise Ratio) as our metric.

**Note:** For training the AOD-Net in a multi-GPU environment, we can also use
`tf.distribute.MirroredStrategy` (go to the
[Distributed training](https://www.tensorflow.org/guide/distributed_training)
guide more information).
```

File: examples/vision/aodnet.py [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+)
+model.build((1, 256, 256, 3))
+
+model.summary()
+
+"""
+# Training
+
+We train AODNet for 10 epochs, on Nvidia Tesla P100, an epoch takes around 190 seconds,
+so the model takes around 30 minutes to train
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:288]:
```suggestion
We train the AOD-Net for 10 epochs. On an NVIDIA Tesla P100, each epoch can take around 190 seconds, so the model may takes around 30 minutes to train.
```

File: examples/vision/aodnet.py [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+plt.title("PSNR Curve")
+plt.ylabel("PSNR")
+plt.xlabel("epoch")
+plt.legend(["train", "val"], loc="upper left")
+plt.show()
+
+"""
+# Inference
+
+We define 2 simple utility functions for inferring from an image and plotting the results
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:320]:
```suggestion
Next, let's define 2 simple utility functions for inferring from an image and plotting 
the results:
```

**Review by 8bitmp3 — COMMENTED** [review]
Another awesome example @soumik12345.

I left a few suggestions, PTAL 👍 hope they help.

**Comment by soumik12345** [comment]
> Another awesome example @soumik12345.
> 
> I left a few suggestions, PTAL 👍 hope they help.

@8bitmp3 Thanks for all the suggestions, updated the example as per your suggestions :)

File: examples/vision/aodnet.py [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+Author: [Soumik Rakshit](https://github.com/soumik12345)
+Date created: 2021/04/10
+Last modified: 2021/04/10
+Description: Dehaze images with a lightweight deep convolutional neural network.
+"""
+"""
+# Introduction
+
+_Image dehazing_ — improving the image quality by removing the haze from images
+captured in real-world weather conditions — is an important problem is computer
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:12]: Oops, a typo!

```suggestion
captured in real-world weather conditions — is an important problem in computer
```

File: examples/vision/aodnet.py [hunk 16]
```diff
@@ -0,0 +1,362 @@
...
+vision. It can be used not only to enhance photographs clicked in hazy
+conditions, but also can be used to augment other computer vision systems, such
+as _object detection_ and _semantic segmentation_ as a
+pre-processing step. In this example, we implement
+the [All-in-One Network (AOD-Net)](https://arxiv.org/abs/1707.06543v1) — a simple and
+lightweight deep convolutional neural network model that can generate clean
+images directly from hazy images. The end-to-end design of the AOD-Net
+enables embedding the network into other neural network models for, such as
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:20]:
```suggestion
enables embedding the network into other neural network models, such as
```

File: examples/vision/aodnet.py [hunk 17]
```diff
@@ -0,0 +1,362 @@
...
+    batch_size=16,
+    val_split=0.1,
+)
+print(train_dataset)
+print(val_dataset)
+
+"""
+## Define the AOD-Net model
+
+Now we will define the AOD-Net model as a subclass of  `tf.keras.Model`. The
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:193]: Nit:

```suggestion
Now we will define the AOD-Net model as a subclass of `tf.keras.Model`. The
```

File: examples/vision/aodnet.py [hunk 18]
```diff
@@ -0,0 +1,362 @@
...
+following diagrams from the original paper summarize its architecture:
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:194]: Suggestion: "its" -> "the model's" (to make sure some users don't get confused)

```suggestion
following diagrams from the original paper summarize the model's architecture:
```

File: examples/vision/aodnet.py [hunk 19]
```diff
@@ -0,0 +1,362 @@
...
+# Training
+
+We train the AOD-Net for 10 epochs. On an NVIDIA Tesla P100, each epoch
+can take around 190 seconds, so the model may take around 30 minutes to train.
+"""
+
+training_history = model.fit(train_dataset, validation_data=val_dataset, epochs=10)
+
+"""
+## Plot of Training and Validation Loss
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:301]: Suggestion: use sentence case, "losses" (plural)

```suggestion
## Plot of training and validation losses
```

File: examples/vision/aodnet.py [hunk 20]
```diff
@@ -0,0 +1,362 @@
...
+plt.plot(training_history.history["loss"])
+plt.plot(training_history.history["val_loss"])
+plt.title("Loss Curve")
+plt.ylabel("loss")
+plt.xlabel("epoch")
+plt.legend(["train", "val"], loc="upper left")
+plt.show()
+
+"""
+## Plot of Training and Validation PSNR
```
Comment (8bitmp3) [inline examples/vision/aodnet.py:313]: Suggestion: use sentence case, "PSNRs" (since it is "ratios", it should be plural)

```suggestion
## Plot of training and validation PSNRs
```

**Comment by fchollet** [comment]
Please let me know when the PR is ready for review again! :)

**Comment by soumik12345** [comment]
> Please let me know when the PR is ready for review again! :)

Sure @fchollet

**Comment by soumik12345** [comment]
@fchollet I attempted to rewrite the augmentations using Keras preprocessing layers, but the problem that I've been unable to solve is the application of exactly the same transform to both the cropped segment of the input image and the corresponding ground truth. Would really appreciate your thoughts on this topic.

**Comment by fchollet** [comment]
> I attempted to rewrite the augmentations using Keras preprocessing layers, but the problem that I've been unable to solve is the application of exactly the same transform to both the cropped segment of the input image and the corresponding ground truth. Would really appreciate your thoughts on this topic.

I would recommend creating 2 identical layers (or sets of layers) with the same `seed` argument in the constructor. They will apply the same transformations as long as they're called in the same order.

**Comment by fchollet** [comment]
Are you still working on this?

**Comment by fchollet** [comment]
Closing PR as stale.
